include/cuda/memory_resource

```diff
@@ -0,0 +1,529 @@
+_LIBCUDACXX_CONCEPT _Has_member_deallocate_async = _LIBCUDACXX_FRAGMENT(_Has_member_deallocate_async_, _Resource);
+
+/// \concept async_resource
+/// \brief The \c async_resource concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT async_resource =
+  resource<_Resource>&& _Has_member_allocate_async<_Resource>&& _Has_member_deallocate_async<_Resource>;
+
+/// \concept has_property
+/// \brief The \c has_property concept
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_property_,
+                             requires(const _Resource& __res) //
+                             (get_property(__res, _Property{})));
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT has_property = _LIBCUDACXX_FRAGMENT(_Has_property_, _Resource, _Property);
+
+template <class T>
+using _Has_value_types = typename T::value_type;
+
+/// \concept property_with_value
+/// \brief The \c property_with_value concept
+template <class _Property>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Property_with_value_,
+                             requires() //
+                             (typename(_Has_value_types<_Property>)));
+template <class _Property>
+_LIBCUDACXX_CONCEPT property_with_value = _LIBCUDACXX_FRAGMENT(_Property_with_value_, _Property);
+
+/// \concept has_property_with
+/// \brief The \c has_property_with concept
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_property_with_,
+                             requires(const _Resource& __res) //
+                             (_CUDA_VSTD::_Returns_exactly<_Return>(get_property(__res, _Property{}))));
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT has_property_with = property_with_value<_Property>&&
+  _LIBCUDACXX_FRAGMENT(_Has_property_with_, _Resource, _Property, _Return);
+
+/// \concept resource_with
+/// \brief The \c resource_with concept
+template <class _Resource, class... _Properties>
+#if _LIBCUDACXX_STD_VER < 17
+_LIBCUDACXX_CONCEPT resource_with =
+  resource<_Resource>&& _CUDA_VSTD::conjunction_v<_CUDA_VSTD::bool_constant<has_property<_Resource, _Properties>>...>;
+#else
+_LIBCUDACXX_CONCEPT resource_with       = resource<_Resource> && (has_property<_Resource, _Properties> && ...);
+#endif
+
+/// \concept async_resource_with
+/// \brief The \c async_resource_with concept
+template <class _Resource, class... _Properties>
+#if _LIBCUDACXX_STD_VER < 17
+_LIBCUDACXX_CONCEPT async_resource_with = async_resource<_Resource>&&
+  _CUDA_VSTD::conjunction_v<_CUDA_VSTD::bool_constant<has_property<_Resource, _Properties>>...>;
+#else
+_LIBCUDACXX_CONCEPT async_resource_with = async_resource<_Resource> //
+                                          && (has_property<_Resource, _Properties> && ...);
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+/// class resource_ref
+/// class async_resource_ref
+enum class _AllocType
+{
+  _Default,
+  _Async,
+};
```

Comment:
I'm not a fan of this enum. Going the enum way and templating basic_resource_ref against it closes the door for the users to use whatever other interfaces they might be intereseted in.

Reply:
I do not see the point here. This is meant as a standard facility. We can happily extend the enum as much as we like. Of course a user should not do that with a standard type.

Note that a user could still extend any `resoure_ref` with arbitrary functionality via `get_property`

Reply:
Yes, but the user would not be able to extend the allocation interface.

Reply:
To be clear.

A user should not under any circumstances expand this type. The whole point of the concepts based approach is to check exactly for the two interfaces.

Consequently, a user cannot modify those. It is just not possible to then fulfil the `resource` and `async_resource` concepts.